controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/pkg/ApplicationPackageValidator.java

```diff
@@ -90,22 +93,40 @@ private void validateSteps(DeploymentSpec deploymentSpec) {
         }
     }
 
-    /** Verify that no single endpoint contains regions in different clouds */
+    /** Verify that:
+     * <ul>
+     *     <li>no single endpoint contains regions in different clouds</li>
+     *     <li>application endpoints with different regions must be contained in CGP and AWS</li>
+     * </ul>
+     */
     private void validateEndpointRegions(DeploymentSpec deploymentSpec) {
         for (var instance : deploymentSpec.instances()) {
-            for (var endpoint : instance.endpoints()) {
-                var clouds = new HashSet<CloudName>();
-                for (var region : endpoint.regions()) {
-                    for (ZoneApi zone : controller.zoneRegistry().zones().all().in(Environment.prod).in(region).zones()) {
-                        clouds.add(zone.getCloudName());
-                    }
-                }
-                if (clouds.size() != 1 && !clouds.equals(Set.of(CloudName.GCP, CloudName.AWS))) {
-                    throw new IllegalArgumentException("Endpoint '" + endpoint.endpointId() + "' in " + instance +
-                                                       " cannot contain regions in different clouds: " +
+            validateEndpointRegions(instance.endpoints(), instance);
+        }
+        validateEndpointRegions(deploymentSpec.endpoints(), null);
+    }
+
+    private void validateEndpointRegions(List<Endpoint> endpoints, DeploymentInstanceSpec instance) {
+        for (var endpoint : endpoints) {
+            RegionName[] regions = new HashSet<>(endpoint.regions()).toArray(RegionName[]::new);
```

Comment:
Why `toArray`?

Reply:
That's the required argument for the `.in(RegionName... regions)`.